txoutsbyaddress index (take 2)

djpnewton opened the pull request:
This is an attempt to revive PR #5048

I have moved the index to its own database and also added a gettxoutsbyaddress REST command

I still need to think about the layout of binary input/output for the REST command, or is bin encoding necessary?

Also still missing tests for the new REST stuff

cozz committed:
txoutsbyaddress index

djpnewton committed:
move txoutsbyaddress index to its own database

djpnewton committed:
add txoutsbyaddress querying via rest interface

djpnewton committed:
C++11: s/boost::scoped_ptr/std::unique_ptr/

dcousens commented:
It's not overly complicated to maintain this index outside of `bitcoind`,   IMHO,  that is where it should be.   Perhaps `indexd` or some other service.

luke-jr commented:
NACK. Txouts are not associated with addresses at all.

paveljanik commented:
@luke-jr They are, but this association is not useful for 99.9% `bitcoind` users. And it slows down the normal work used by 100% of `bitcoind` users. So NACK.

If people need this, please contribute some description/work using the external indexing solution into `contrib/`.

btcdrak commented:
Concept ACK from me.
@paveljanik The feature is off by default so it shouldn't slow down anything for normal users.

MarcoFalke commented on qa/rpc-tests/txoutsbyaddress.sh:
```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env bash
```
We don't use bash anymore in /rpc-tests

laanwj replied:
Yes, please look at the existing tests in `qa/rpc-tests`, use the Python framework. Also add the test to `qa/pull-tester/rpc-tests.py` so it gets executed by default.

djpnewton commented:
The inspiration for this PR is to try and make it easier to build a fully validating wallet that does not use significantly more resources then a full node itself.

It seems that most (all?) current solutions for building a wallet either involve added trust (SPV or centralized API services) or added resource requirements (bitcore.io needs 8GB RAM and 200GB disk space)

Thats the rationale (which I probably should have stated at the outset) anyway. The last time I was trying to make a demo it was easier just to use the blockcypher API, or maybe there are projects that I missed that fill this gap?

luke-jr commented:
If the RPC was by script (as the internals seem to be) instead of by address, it might make sense.

dcousens commented:
> If people need this, please contribute some description/work using the external indexing solution into contrib/.

> OT: If indexes are really considered so bad then -txindex should be removed.

IMHO,  that is where `txindex` it should be,  at least then it might easy to extend to things like this without adding 1000 SLOC.

djpnewton commented:
@luke-jr the RPC accepts both hex encoded script or addresses

jonasschnelli commented:
> The inspiration for this PR is to try and make it easier to build a fully validating wallet that does not use significantly more resources then a full node itself.

> It seems that most (all?) current solutions for building a wallet either involve added trust (SPV or centralized API services) or added resource requirements (bitcore.io needs 8GB RAM and 200GB disk space)

These are good point!

Some thoughts:
1) If we would add something like this, it should be more modular. Have a look at https://github.com/bitcoin/bitcoin/pull/8501/files, there you can see how just a handful of lines in `init.cpp` can add core features over a flexible modular approach. It also simplifies rebases.

2) Was it considered to add the index outside of bitcoin-core? I guess using ZMQ notifications together with REST (or RPC) would work. What would be the downsides or advantages of the "external" approach?

dcousens commented:
> Was it considered to add the index outside of bitcoin-core? I guess using ZMQ notifications together with REST (or RPC) would work. What would be the downsides or advantages of the "external" approach?

As someone who does this approach,   the only difficulties is ensuring that you handle orphans and forks properly.
They do happen enough and cause issues,  so you need to be prepared for the floor to fall out between RPC calls.
Simple enough code,  but battle testing it can take a while and would probably be important to have examples for in `contrib/` for beginners.

btcdrak commented:
Indexes are desperately needed, there are quite a few users of my addrindex fork https://github.com/btcdrak/bitcoin/releases that demonstrate this need and there is a steady flow of people finding out about it. This would be a more lightweight option and I think it should be part of Bitcoin Core in an easily accessible way (i.e. just turn on with a switch+reindex). While I could say add it to my fork, it is does reduce the visibility of it.

sipa commented:
`-txindex` was only added for backward compatibility with the pre-0.8 `getrawtransaction` RPC. I think people who need it should build an external solution. It's inefficient, requires updates all over the place, and doesn't even guarantee consistency when there are reorgs, and is not needed for any of the functions Bitcoin Core provides. The addrindex hack is even worse - it encourages people to build an external wallet in an easy, unscalable, and inadvisable way.

Indexing the UTXO set like this PR does is less bad - at least it doesn't prevent pruning - but what function would it provide apart from RPCs? The wallet can't actually use it (as it would miss historical transactions), and it won't be available in SPV mode.

jonasschnelli commented:
Agree with @sipa.

A maybe more scalable way how "wallet" (or lets say something like the BWS [bitpay wallet service]) could be built on top of bitcoin core would be:
* add multiwallet support (something similar like https://github.com/bitcoin/bitcoin/pull/2184)
* add support for pubkey only wallets (should be relatively trivial, partially solved with the current support for "watchonly")
* external wallet could use RPC in order to create "wallets", use `fundrawtransaction` in a pubkey-only multiwallet setup and 
* transactions can and should be signed in the "external wallet space"

"Wallets" would then be reduces to...
* A set of addresses to keep track of (track/store transactions)
* Coinselection and utxo per set

Restoring backups is the only feature I can think off that would require a addr-index (otherwise rescanning will be very slow).

djpnewton commented:
The entire point of running a node is to provide validation for a wallet somewhere. That is what makes a node significant.

Is it actually simple to provide this service outside of bitcoin-core? Are there any projects that manage to do it without duplicating the entire blockchain into their own database? It seems a real shame if we go to the trouble of assembling the UTXO set and cant provide ways for external wallets to leverage it.

@sipa are you sure it would miss historical transactions? Any transactions adding funds to the wallet will show up in the UTXO set and any transactions spending funds from the wallet can only be created by the wallet itself (I guess they can be malleated).

@jonasschnelli I think I like this idea. It might not be as flexible in the kinds of external wallets it allows though.

dcousens commented:
> Are there any projects that manage to do it without duplicating the entire blockchain into their own database? It seems a real shame if we go to the trouble of assembling the UTXO set and cant provide ways for external wallets to leverage it.

Of course,  but this index is not small by any means.

laanwj commented:
I think this functionality is very useful. E.g. to get the current balance for an arbitrary address/script.
Missed this a few times. I think external wallets such as `coinjoin` may find this useful as well.

- **Sure**, you cannot use this with SPV but the primary reason for Bitcoin Core's existence is to be a full node. Storing the UTXO set comes with that. To be clear SPV mode is by no means the future of this project, at most it would be optional.

- Unlike a full tx index it can be used with pruning, and takes a lot less disk space.

Concept ACK.

laanwj commented:
>  is bin encoding necessary?

No it's not necessary - only if there is a straightforward binary representation like for transactions and blocks.

dcousens commented:
I wasn't aware this was just for UTXOs,  not txouts *at any point* in history (aka, a full txout index). 

concept ACK

dexX7 commented:
Concept ACK.

Several projects I know of use the addrindex stuff from years ago to derive information about UTXO state, and this is much more elegant.

dexX7 commented on src/init.cpp:
```diff
@@ -340,6 +345,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-sysperms", _("Create new files with system default permissions, instead of umask 077 (only effective with disabled wallet functionality)"));
 #endif
     strUsage += HelpMessageOpt("-txindex", strprintf(_("Maintain a full transaction index, used by the getrawtransaction rpc call (default: %u)"), DEFAULT_TXINDEX));
+    strUsage += HelpMessageOpt("-txoutsbyaddressindex", strprintf(_("Maintain an address to unspent outputs index (rpc: gettxoutsbyaddress). The index is built on first use. (default: %u)"), 0));
```
My personal preference would be something shorter like `"txoutsindex"`, especially because it really indexes scripts and not addresses.

djpnewton replied:
Sure I dont see a problem with that

MarcoFalke replied:
Or just `txoutindex`

djpnewton replied:
argh, paradox of choice!

Ok, how about if nobody has any objection I will change the name to "txoutindex" (command line parameter, RPC, REST etc)?

djpnewton committed:
convert txoutsbyaddress test to python

djpnewton committed:
- add txoutsbyaddress REST interface tests
 - remove skeleton HEX/BIN format support from REST interface for gettxoutsbyaddress endpoint

djpnewton commented:
Ok I converted the RPC tests to python and added some REST tests

MarcoFalke commented on qa/rpc-tests/rest.py:
```diff
@@ -53,7 +53,8 @@ def __init__(self):
 
     def setup_network(self, split=False):
-        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir)
+        args = ["-txoutsbyaddressindex"]
```
Nit: You can put something like the following up a few lines, so the scope is not limited to `setup_network`.

```py
self.extra_args = [["-txoutsbyaddressindex"]] * 3

MarcoFalke commented on qa/rpc-tests/txoutsbyaddress.py:
```diff
@@ -0,0 +1,96 @@
+        self.nodes.append(start_node(1, self.options.tmpdir))
+        self.nodes.append(start_node(2, self.options.tmpdir))
+        self.is_network_split = False
+        self.sync_all()
```
Nit: I think this is not required

MarcoFalke commented on qa/rpc-tests/txoutsbyaddress.py:
```diff
@@ -0,0 +1,96 @@
+
+    def setup_network(self, split=False):
+        print("Setup network...")
+        self.nodes = []
```
```py
self.nodes = start_nodes(3, self.options.tmpdir, self.extra_args)
```
should work as well. (You can put `self.extra_args = [['-txoutindex'], []. []]` in `__init__()`)

MarcoFalke commented on qa/rpc-tests/txoutsbyaddress.py:
```diff
@@ -0,0 +1,96 @@
+            assert_equal(len(node.listunspent()), 0)
+
+        # mining
+        connect_nodes(self.nodes[1], 0)
```
Nit: you can move this up into `setup_network()`. I think all other tests do this

MarcoFalke commented on qa/rpc-tests/txoutsbyaddress.py:
```diff
@@ -0,0 +1,96 @@
+        assert_is_hash_string(txid)
+        assert_equal(txid, txid1)
+
+        # stop node 2
```
Why is this required?

djpnewton replied:
node 2 is stopped so that later on it can be restarted and then orphan a block created by node 1 (to test txoutindex handles the reorg)

basically its stopped so I will not build on a block created by node 1

ryanofsky replied:
Should probably mention this in a comment.

djpnewton committed:
address test review nits

djpnewton commented:
addressed review nits by @MarcoFalke (thank you)

jmcorgan commented:
Concept ACK.

MarcoFalke commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -143,6 +143,7 @@
     'importprunedfunds.py',
     'signmessages.py',
     'p2p-compactblocks.py',
+    'txoutsbyaddress.py',
```
Missing `+x` permission

djpnewton replied:
done

djpnewton committed:
rename -txoutsbyaddress to -txoutindex

djpnewton committed:
set executable bit

djpnewton committed:
add doc for gettxoutsbyaddress REST endpoint

btcdrak commented:
needs rebase

gmaxwell commented:
Concept ACK. Comment, can others provide feeback: I think the argument should be changed to unspentoutputindex or utxoindex or something like that, otherwise it sounds like it indexes everything. Similar for the rpc names.

droark commented:
Hmmm. If @djpnewton doesn't pick this up again in the next few days, I'll seriously consider keeping it going.

ryanofsky commented on src/coinsbyscript.h:
```diff
@@ -0,0 +1,63 @@
+    }
+};
+
+typedef std::map<uint160, CCoinsByScript> CCoinsMapByScript; // uint160 = hash of script
```
Maybe replace uint160 with CScriptID to make the map type more obvious.

ryanofsky commented on qa/rpc-tests/txoutsbyaddress.py:
```diff
@@ -0,0 +1,93 @@
+        self.sync_all()
+        assert_equal(self.nodes[0].getbalance(), 5090)
+        assert_equal(self.nodes[1].getbalance(), 10)
+        txouts = self.nodes[0].gettxoutsbyaddress(1, (address,))
```
Maybe check other txouts fields here as well.

ryanofsky commented on src/coinsbyscript.cpp:
```diff
@@ -0,0 +1,278 @@
+    if (!hashBlock.IsNull())
+        batch.Write(DB_BEST_BLOCK, hashBlock);
+
+    LogPrint("coindb", "Committing %u coin address indexes to coin database...\n", (unsigned int)count);
```
Could use %z instead of (unsigned int) cast, I think.

ryanofsky reviewed (commented):
Reviewed most of this PR, but not the whole thing. Mostly minor comments, except for a possible bug in CCoinsViewByScript::FetchCoinsByScript.

paveljanik commented:
Needs rebase.

djpnewton committed:
fix up a change missed by the rebase

MarcoFalke commented:
> I am thinking about unifying it around "-utxoindex"

Sounds good.

btcdrak commented:
needs rebase again.

ryanofsky commented:
@djpnewton or @droark have you done more work in this? I'd be happy to help continue the effort on this feature.

droark commented:
@ryanofsky - I hadn't been working on this but I'm happy to run with the ball since the OP seems to have disappeared again. I'll post a new PR once it's up and running.

Thanks.

fanquake commented:
Closing this for now then. @droak or @ryanofsky can open a new PR once they've restarted the work.
